src/lib.rs

```diff
@@ -64,18 +64,42 @@ impl<T> Arena<T> {
 
     /// Uses the contents of an iterator to allocate values in the arena.
     /// Returns a mutable slice that contains these values.
-    pub fn alloc_extend<I>(&self, iter: I) -> &mut [T]
-        where I: Iterator<Item = T> + ExactSizeIterator
+    pub fn alloc_extend<I>(&self, iterable: I) -> &mut [T]
+        where I: IntoIterator<Item = T>
     {
+        let mut iter = iterable.into_iter();
+
         let mut chunks = self.chunks.borrow_mut();
 
-        if chunks.current.len() + iter.len() > chunks.current.capacity() {
-            chunks.reserve(iter.len());
+        let iter_min_len = iter.size_hint().0;
+        let mut next_item_index;
+        if chunks.current.len() + iter_min_len > chunks.current.capacity() {
+            chunks.reserve(iter_min_len);
+            chunks.current.extend(iter);
+            next_item_index = 0;
+        } else {
+            next_item_index = chunks.current.len();
+            let mut i = 0;
+            while let Some(elem) = iter.next() {
+                if i == iter_min_len {
+                    // The iterator were larger than its `size_hint` indicated.
+                    let chunks = &mut *chunks;
+                    // Create a new chunk which we can freely push the entire iterator into
+                    chunks.reserve(1);
```

Comment:
Not really sure how inlining or creating a new method would help. `reserve` does exactly what we want in this case. It _reserves_ an additional chunk which has room for at least one more element (I should actually be passing `iter_min_len + 1` instead since `reserve` takes the minimum size of the next chunk, that's only a matter of avoiding reallocations though).